File: ticket-be/src/main/java/com/ticket/service/AuthService.java
package com.ticket.service;

import com.ticket.dto.JwtResponse;
import com.ticket.dto.LoginRequest;
import com.ticket.dto.RegisterRequest;
import com.ticket.entity.Role;
import com.ticket.entity.Role.RoleName;
import com.ticket.entity.User;
import com.ticket.repository.RoleRepository;
import com.ticket.repository.UserRepository;
import com.ticket.security.JwtUtils;
import com.ticket.security.UserDetailsImpl;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
public class AuthService {
    private static final Logger logger = LoggerFactory.getLogger(AuthService.class);

    @Autowired
    private AuthenticationManager authenticationManager;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private RoleRepository roleRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Autowired
    private JwtUtils jwtUtils;

    @Autowired
    private RefreshTokenService refreshTokenService;

    @Autowired
    private TokenBlacklistService tokenBlacklistService;

    @Transactional
    public JwtResponse login(LoginRequest loginRequest) {
        Authentication authentication = authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(loginRequest.getUsername(), loginRequest.getPassword())
        );

        SecurityContextHolder.getContext().setAuthentication(authentication);
        
        UserDetailsImpl userDetails = (UserDetailsImpl) authentication.getPrincipal();
        String accessToken = jwtUtils.generateAccessToken(authentication);
        String refreshToken = refreshTokenService.createRefreshToken(userDetails.getId());
        
        List<String> roles = userDetails.getAuthorities().stream()
                .map(item -> item.getAuthority())
                .collect(Collectors.toList());

        logger.info("User {} logged in successfully", userDetails.getUsername());

        return new JwtResponse(
                accessToken,
                refreshToken,
                userDetails.getId(),
                userDetails.getUsername(),
                userDetails.getEmail(),
                roles
        );
    }

    @Transactional
    public JwtResponse refreshToken(String refreshToken) {
        UUID userId = refreshTokenService.verifyRefreshToken(refreshToken);
        
        if (userId == null) {
            throw new RuntimeException("Refresh token không hợp lệ hoặc đã hết hạn!");
        }
        
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User không tồn tại!"));
        
        String newAccessToken = jwtUtils.generateAccessTokenFromUsername(user.getUsername());
        String newRefreshToken = refreshTokenService.rotateRefreshToken(refreshToken, userId);
        
        List<String> roles = user.getRoles().stream()
                .map(role -> role.getName().name())
                .collect(Collectors.toList());

        logger.info("Token refreshed for user {}", user.getUsername());

        return new JwtResponse(
                newAccessToken,
                newRefreshToken,
                user.getId(),
                user.getUsername(),
                user.getEmail(),
                roles
        );
    }

    @Transactional
    public void logout(String accessToken, String refreshToken, boolean logoutAll) {
        if (accessToken != null) {
            long ttl = jwtUtils.getRemainingTimeInSeconds(accessToken);
            if (ttl > 0) {
                tokenBlacklistService.addToBlacklist(accessToken, ttl);
                logger.info("Access token added to blacklist with TTL: {} seconds", ttl);
            }
        }
        
        if (refreshToken != null) {
            if (logoutAll) {
                UUID userId = refreshTokenService.verifyRefreshToken(refreshToken);
                if (userId != null) {
                    refreshTokenService.deleteAllRefreshTokensOfUser(userId);
                    logger.info("User {} logged out from all devices", userId);
                }
            } else {
                refreshTokenService.deleteRefreshToken(refreshToken);
                logger.info("Refresh token deleted");
            }
        }
    }

    @Transactional
    public void logout(String accessToken) {
        logout(accessToken, null, false);
    }

    @Transactional
    public void registerCustomer(RegisterRequest registerRequest) {
        validateRegisterRequest(registerRequest);
        User user = createUser(registerRequest);
        
        Role customerRole = roleRepository.findByName(RoleName.ROLE_CUSTOMER)
                .orElseThrow(() -> new RuntimeException("Error: Role CUSTOMER is not found."));
        
        Set<Role> roles = new HashSet<>();
        roles.add(customerRole);
        user.setRoles(roles);

        userRepository.save(user);
    }

    @Transactional
    public void registerOrganizer(RegisterRequest registerRequest) {
        validateRegisterRequest(registerRequest);
        User user = createUser(registerRequest);
        
        Role organizerRole = roleRepository.findByName(RoleName.ROLE_ORGANIZER)
                .orElseThrow(() -> new RuntimeException("Error: Role ORGANIZER is not found."));
        
        Set<Role> roles = new HashSet<>();
        roles.add(organizerRole);
        user.setRoles(roles);

        userRepository.save(user);
    }

    @Transactional
    public void registerAdmin(RegisterRequest registerRequest) {
        validateRegisterRequest(registerRequest);
        User user = createUser(registerRequest);
        
        Role adminRole = roleRepository.findByName(RoleName.ROLE_ADMIN)
                .orElseThrow(() -> new RuntimeException("Error: Role ADMIN is not found."));
        
        Set<Role> roles = new HashSet<>();
        roles.add(adminRole);
        user.setRoles(roles);

        userRepository.save(user);
    }

    private void validateRegisterRequest(RegisterRequest registerRequest) {
        if (userRepository.existsByUsername(registerRequest.getUsername())) {
            throw new RuntimeException("Error: Username đã được sử dụng!");
        }
        if (userRepository.existsByEmail(registerRequest.getEmail())) {
            throw new RuntimeException("Error: Email đã được sử dụng!");
        }
    }

    private User createUser(RegisterRequest registerRequest) {
        User user = new User();
        user.setUsername(registerRequest.getUsername());
        user.setEmail(registerRequest.getEmail());
        user.setPassword(passwordEncoder.encode(registerRequest.getPassword()));
        user.setFullName(registerRequest.getFullName());
        user.setPhoneNumber(registerRequest.getPhoneNumber());
        user.setActive(true);
        return user;
    }
}

File: ticket-fe/apis/auth.service.ts
import api from "@/config/axios.config";

// =============== REQUEST DTOs ===============

export interface LoginRequest {
  username: string;
  password: string;
}

export interface RegisterRequest {
  username: string;
  email: string;
  password: string;
  fullName: string;
  phoneNumber?: string;
}

export interface RefreshTokenRequest {
  refreshToken: string;
}

export interface LogoutRequest {
  refreshToken?: string;
  logoutAll?: boolean;
}

// =============== RESPONSE DTOs ===============

export interface JwtResponse {
  accessToken: string;
  refreshToken: string;
  id: string;
  username: string;
  email: string;
  roles: string[];
}

export interface MessageResponse {
  message: string;
}

// =============== AUTH SERVICE ===============

export const authService = {
  /**
   * Đăng nhập
   * POST /auth/login
   */
  async login(request: LoginRequest): Promise<JwtResponse> {
    const response = await api.post<JwtResponse>("/auth/login", request);
    
    // Lưu tokens vào localStorage
    if (response.data.accessToken) {
      localStorage.setItem("accessToken", response.data.accessToken);
      localStorage.setItem("refreshToken", response.data.refreshToken);
      localStorage.setItem("user", JSON.stringify({
        id: response.data.id,
        username: response.data.username,
        email: response.data.email,
        roles: response.data.roles,
      }));
    }
    
    return response.data;
  },

  /**
   * Đăng ký tài khoản Khách hàng (CUSTOMER)
   * POST /auth/register/customer
   */
  async registerCustomer(request: RegisterRequest): Promise<MessageResponse> {
    const response = await api.post<MessageResponse>("/auth/register/customer", request);
    return response.data;
  },

  /**
   * Đăng ký tài khoản Nhà tổ chức (ORGANIZER)
   * POST /auth/register/organizer
   */
  async registerOrganizer(request: RegisterRequest): Promise<MessageResponse> {
    const response = await api.post<MessageResponse>("/auth/register/organizer", request);
    return response.data;
  },

  /**
   * Làm mới Access Token
   * POST /auth/refresh-token
   */
  async refreshToken(): Promise<JwtResponse> {
    const refreshToken = localStorage.getItem("refreshToken");
    
    if (!refreshToken) {
      throw new Error("No refresh token available");
    }
    
    const response = await api.post<JwtResponse>("/auth/refresh-token", {
      refreshToken,
    });
    
    // Cập nhật tokens mới
    if (response.data.accessToken) {
      localStorage.setItem("accessToken", response.data.accessToken);
      localStorage.setItem("refreshToken", response.data.refreshToken);
    }
    
    return response.data;
  },

  /**
   * Đăng xuất
   * POST /auth/logout
   */
  async logout(logoutAll: boolean = false): Promise<void> {
    try {
      const refreshToken = localStorage.getItem("refreshToken");
      
      await api.post<MessageResponse>("/auth/logout", {
        refreshToken,
        logoutAll,
      });
    } finally {
      // Xóa dữ liệu đăng nhập khỏi localStorage
      localStorage.removeItem("accessToken");
      localStorage.removeItem("refreshToken");
      localStorage.removeItem("user");
    }
  },

  /**
   * Kiểm tra đã đăng nhập chưa
   */
  isAuthenticated(): boolean {
    if (typeof window === "undefined") return false;
    return !!localStorage.getItem("accessToken");
  },

  /**
   * Lấy thông tin user từ localStorage
   */
  getCurrentUser(): { id: string; username: string; email: string; roles: string[] } | null {
    if (typeof window === "undefined") return null;
    
    const userStr = localStorage.getItem("user");
    if (!userStr) return null;
    
    try {
      return JSON.parse(userStr);
    } catch {
      return null;
    }
  },

  /**
   * Kiểm tra user có role cụ thể
   */
  hasRole(role: string): boolean {
    const user = this.getCurrentUser();
    if (!user) return false;
    return user.roles.includes(role);
  },

  /**
   * Kiểm tra user là Customer
   */
  isCustomer(): boolean {
    return this.hasRole("ROLE_CUSTOMER");
  },

  /**
   * Kiểm tra user là Organizer
   */
  isOrganizer(): boolean {
    return this.hasRole("ROLE_ORGANIZER");
  },

  /**
   * Kiểm tra user là Admin
   */
  isAdmin(): boolean {
    return this.hasRole("ROLE_ADMIN");
  },
};

export default authService;

